android/src/main/java/com/tailscale/ipn/IPNService.kt

```diff
@@ -0,0 +1,130 @@
+// Copyright (c) Tailscale Inc & AUTHORS
+// SPDX-License-Identifier: BSD-3-Clause
+package com.tailscale.ipn
+
+import android.app.PendingIntent
+import android.content.Intent
+import android.content.pm.PackageManager
+import android.net.VpnService
+import android.os.Build
+import android.system.OsConstants
+import androidx.core.app.NotificationCompat
+import androidx.core.app.NotificationManagerCompat
+import libtailscale.Libtailscale
+import java.util.UUID
+
+open class IPNService : VpnService(), libtailscale.IPNService {
+    private val randomID: String = UUID.randomUUID().toString()
+    override fun id(): String {
+        return randomID
+    }
+
+    override fun onStartCommand(intent: Intent, flags: Int, startId: Int): Int {
+        if (intent != null && ACTION_STOP_VPN == intent.getAction()) {
+            (getApplicationContext() as App).autoConnect = false
+            close()
+            return START_NOT_STICKY
+        }
+        val app = getApplicationContext() as App
+        if (intent != null && "android.net.VpnService" == intent.getAction()) {
+            // Start VPN and connect to it due to Always-on VPN
+            val i = Intent(IPNReceiver.INTENT_CONNECT_VPN)
+            i.setPackage(getPackageName())
+            i.setClass(getApplicationContext(), IPNReceiver::class.java)
+            sendBroadcast(i)
+            Libtailscale.requestVPN(this)
+            app.connect()
+            return START_STICKY
+        }
+        Libtailscale.requestVPN(this)
+        if (app.vpnReady && app.autoConnect) {
+            app.connect();
+        }
+        return START_STICKY
+    }
+
+
+    private fun close() {
+        stopForeground(true)
+        Libtailscale.serviceDisconnect(this)
+    }
+
+    override fun onDestroy() {
+        close()
+        super.onDestroy()
+    }
+
+    override fun onRevoke() {
+        close()
+        super.onRevoke()
+    }
+
+    private fun configIntent(): PendingIntent {
+        return PendingIntent.getActivity(this, 0, Intent(this, IPNActivity::class.java),
+                PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE)
+    }
+
+    private fun disallowApp(b: Builder, name: String) {
+        try {
+            b.addDisallowedApplication(name)
+        } catch (e: PackageManager.NameNotFoundException) {
+        }
+    }
+
+    override fun newBuilder(): VPNServiceBuilder {
+        val b: Builder = Builder()
+                .setConfigureIntent(configIntent())
+                .allowFamily(OsConstants.AF_INET)
+                .allowFamily(OsConstants.AF_INET6)
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) b.setMetered(false) // Inherit the metered status from the underlying networks.
+        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M) b.setUnderlyingNetworks(null) // Use all available networks.
+
+        // RCS/Jibe https://github.com/tailscale/tailscale/issues/2322
```

Comment:
These should probably be listed in a manifest somewhere - maybe just a file we include in the apk for now.

Really, we should fetch this list, and apply it dynamically since it seems like it grows over time but we can save that for another day.